File: spec/src/authoring.test.ts
import { describe, expect, it } from 'vitest'
import { AuthoringError, component, navItem, newPage, refresh, setState } from './authoring.js'
import { API_VERSION, KIND_PAGE } from './types.js'

// TS counterpart of authoring/authoring_test.go — keep behaviors aligned.

describe('authoring builders', () => {
  it('builds a dashboard page with bindings, theme, and interactions', () => {
    const page = newPage('sales', 'Sales Dashboard')
      .profile('dashboard')
      .context('customerId', 'cust-42')
      .grid(12, '16px')
      .theme('brand', { primary: '#0f766e', radius: '0.5rem' })
      .add(
        component('revenue', 'analytics.metric')
          .at(0, 0)
          .span(3, 1)
          .prop('title', 'Revenue')
          .bind('primary', 'sales-data', 'totalRevenue', {
            customerId: '${context.customerId}',
          })
          .default('primary', 0),
        component('note', 'core.text')
          .prop('content', 'Filtered')
          .visibleWhen('${state.filters.active}'),
      )
      .onEvent(
        'revenue',
        'click',
        setState('revenue', 'filters.month', '${event.label}'),
        refresh('revenue'),
      )
      .build()

    expect(page.apiVersion).toBe(API_VERSION)
    expect(page.kind).toBe(KIND_PAGE)
    expect(page.metadata.name).toBe('sales')
    expect(page.layout).toEqual({ type: 'responsive-grid', config: { columns: 12, gap: '16px' } })
    expect(page.components[0].position).toEqual({ row: 0, col: 0, colSpan: 3, rowSpan: 1 })
    expect(page.components[0].data?.primary).toEqual({
      source: 'sales-data',
      operation: 'totalRevenue',
      parameters: { customerId: '${context.customerId}' },
      default: 0,
    })
    expect(page.interactions?.[0].then).toHaveLength(2)
  })

  it('builds app shells with nested navigation and children', () => {
    const page = newPage('record', 'Record')
      .appShell('header', 'sidebar', 'main')
      .navigation(
        'sidebar',
        navItem('overview', 'Overview', '/overview'),
        navItem('records', 'Records', undefined, navItem('accounts', 'Accounts', '/accounts')),
      )
      .add(
        component('card', 'core.card')
          .prop('title', 'Acme')
          .child(component('detail', 'application.record-detail').bindStatic('record', { a: 1 })),
      )
      .build()

    expect(page.layout.regions?.map((r) => r.name)).toEqual(['header', 'sidebar', 'main'])
    expect(page.navigation?.items[1].children?.[0].id).toBe('accounts')
    expect(page.components[0].children).toHaveLength(1)
  })

  it('bindState and bindStatic produce the canonical binding shapes', () => {
    const page = newPage('p', 'P')
      .stack()
      .add(component('f', 'analytics.filter').bindState('value', 'filters.period', 'monthly'))
      .build()
    expect(page.components[0].data?.value).toEqual({
      source: 'state',
      operation: 'get',
      parameters: { path: 'filters.period' },
      default: 'monthly',
    })
  })

  it('reports every problem at build time', () => {
    const builder = newPage('', 'No ID').add(component('', '').default('missing', 1))
    let err: AuthoringError | undefined
    try {
      builder.build()
    } catch (e) {
      err = e as AuthoringError
    }
    expect(err).toBeInstanceOf(AuthoringError)
    expect(err!.problems.join('; ')).toContain('page id is required')
    expect(err!.problems.join('; ')).toContain('component id is required')
    expect(err!.problems.join('; ')).toContain('before any binding')
    expect(err!.problems.join('; ')).toContain('a layout is required')
  })

  it('build results are isolated from later mutations', () => {
    const b = newPage('p', 'P').stack()
    const first = b.build()
    b.add(component('late', 'core.text').prop('content', 'x'))
    expect(first.components).toHaveLength(0)
  })
})

File: authoring/authoring_test.go
package authoring_test

import (
	"strings"
	"testing"

	"github.com/plexusone/uiforge/authoring"
	"github.com/plexusone/uiforge/registry"
	"github.com/plexusone/uiforge/uispec"
)

// TestBuildDashboardValidatesAgainstRegistry builds a representative
// dashboard fluently and proves the result passes full registry validation —
// the builder's whole reason to exist.
func TestBuildDashboardValidatesAgainstRegistry(t *testing.T) {
	page, err := authoring.NewPage("sales", "Sales Dashboard").
		Profile(uispec.ProfileDashboard).
		Description("Built by the authoring package").
		Context("customerId", "cust-42").
		Grid(12, "16px").
		Theme("brand", map[string]string{"primary": "#0f766e", "radius": "0.5rem"}).
		Add(
			authoring.Component("revenue", "analytics.metric").
				At(0, 0).Span(3, 1).
				Prop("title", "Revenue").
				Bind("primary", "sales-data", "totalRevenue",
					authoring.Param("customerId", "${context.customerId}")).
				Default("primary", 0),
			authoring.Component("trend", "analytics.line-chart").
				At(0, 3).Span(9, 1).
				Prop("title", "Trend").
				Bind("primary", "sales-data", "revenueByMonth"),
			authoring.Component("note", "core.text").
				Prop("content", "Filtered").
				VisibleWhen("${state.filters.active}"),
		).
		OnEvent("trend", "click",
			authoring.SetState("trend", "filters.month", "${event.label}"),
			authoring.Refresh("revenue"),
		).
		Build()
	if err != nil {
		t.Fatal(err)
	}

	if page.APIVersion != uispec.APIVersion || page.Kind != uispec.KindPage {
		t.Errorf("apiVersion/kind not set: %+v", page)
	}
	if page.Metadata.Name != "sales" {
		t.Errorf("metadata.name should default to id, got %q", page.Metadata.Name)
	}

	r, err := registry.NewWithBuiltins()
	if err != nil {
		t.Fatal(err)
	}
	if err := r.ValidatePage(page); err != nil {
		t.Fatalf("built page should pass registry validation: %v", err)
	}
}

func TestBuildAppShellWithNavigationAndChildren(t *testing.T) {
	page, err := authoring.NewPage("record", "Customer Record").
		Profile(uispec.ProfileApplication).
		AppShell("header", "sidebar", "main").
		Navigation("sidebar",
			authoring.NavItem("overview", "Overview", "/overview"),
			authoring.NavItem("records", "Records", "",
				authoring.NavItem("accounts", "Accounts", "/accounts"),
			),
		).
		Add(
			authoring.Component("card", "core.card").
				Prop("title", "Acme Corp").
				Child(
					authoring.Component("detail", "application.record-detail").
						Prop("title", "Details").
						Bind("record", "crm", "getCustomer").
						Default("record", map[string]any{"owner": "J. Doe"}),
				),
		).
		Build()
	if err != nil {
		t.Fatal(err)
	}

	if page.Navigation == nil || page.Navigation.Items[1].Children[0].ID != "accounts" {
		t.Errorf("nested navigation not built: %+v", page.Navigation)
	}
	if len(page.Components[0].Children) != 1 {
		t.Fatalf("card should have one child")
	}

	r, _ := registry.NewWithBuiltins()
	if err := r.ValidatePage(page); err != nil {
		t.Fatalf("app-shell page should validate: %v", err)
	}
}

func TestBindStateAndStatic(t *testing.T) {
	c := authoring.Component("f", "analytics.filter").
		Prop("label", "Period").Prop("field", "period").Prop("filterType", "select").
		BindState("value", "filters.period", "monthly")
	page, err := authoring.NewPage("p", "P").Stack("vertical").Add(c).Build()
	if err != nil {
		t.Fatal(err)
	}
	b := page.Components[0].Data["value"]
	if b.Source != "state" || b.Parameters["path"] != "filters.period" || b.Default != "monthly" {
		t.Errorf("BindState wrong: %+v", b)
	}

	page2, _ := authoring.NewPage("p2", "P2").Stack("vertical").Add(
		authoring.Component("t", "analytics.table").
			BindStatic("primary", []any{map[string]any{"a": 1}}),
	).Build()
	b2 := page2.Components[0].Data["primary"]
	if b2.Source != "static" || b2.Operation != "value" {
		t.Errorf("BindStatic wrong: %+v", b2)
	}
}

func TestBuildReportsProblems(t *testing.T) {
	_, err := authoring.NewPage("", "No ID").
		Theme("brand", map[string]string{"color-blurple": "#123"}).
		Add(authoring.Component("", "")).
		Build()
	if err == nil {
		t.Fatal("expected build errors")
	}
	for _, want := range []string{
		"page id is required",
		`theme token "color-blurple"`,
		"component id is required",
		"layout is required",
	} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("error should mention %q, got: %v", want, err)
		}
	}

	_, err = authoring.NewPage("p", "P").Stack("vertical").Add(
		authoring.Component("x", "core.text").Default("missing", 1),
	).Build()
	if err == nil || !strings.Contains(err.Error(), "before any binding") {
		t.Errorf("Default without binding should error, got: %v", err)
	}
}

func TestBuilderIsReusableAndIsolated(t *testing.T) {
	b := authoring.NewPage("p", "P").Stack("vertical")
	first, err := b.Build()
	if err != nil {
		t.Fatal(err)
	}
	b.Add(authoring.Component("late", "core.text").Prop("content", "x"))
	if len(first.Components) != 0 {
		t.Error("Build result should be isolated from later mutations")
	}
}
